pulp_ansible/app/downloaders.py

```diff
@@ -0,0 +1,83 @@
+from logging import getLogger
+import asyncio
+import backoff
+import json
+
+from aiohttp.client_exceptions import ClientResponseError
+
+from pulpcore.plugin.download import http_giveup, HttpDownloader
+
+
+log = getLogger(__name__)
+
+
+class TokenAuthHttpDownloader(HttpDownloader):
+    """
+    Custom Downloader that automatically handles Token Based and Basic Authentication.
+
+    """
+
+    token_lock = asyncio.Lock()
+
+    def __init__(self, *args, **kwargs):
+        """
+        Initialize the downloader.
+        """
+        self.remote = kwargs.pop("remote")
+        super().__init__(*args, **kwargs)
+
+    @backoff.on_exception(backoff.expo, ClientResponseError, max_tries=10, giveup=http_giveup)
+    async def _run(self, extra_data=None):
+        """
+        Download, validate, and compute digests on the `url`. This is a coroutine.
+
+        This method is decorated with a backoff-and-retry behavior to retry HTTP 429 errors. It
+        retries with exponential backoff 10 times before allowing a final exception to be raised.
+
+        This method provides the same return object type and documented in
+        :meth:`~pulpcore.plugin.download.BaseDownloader._run`.
+
+        Ansible token reference:
+            https://github.com/ansible/ansible/blob/devel/lib/ansible/galaxy/token.py
+
+        """
+        if not self.remote.token:
+            # No Token
+            return await super()._run(extra_data=extra_data)
+
+        if not self.remote.auth_url:
+            # Galaxy Token
+            headers = {"Authorization": self.remote.token}
+        else:
+            token = await self.update_token_from_auth_url()
+            # Keycloak Token
+            headers = {"Authorization": "Bearer {token}".format(token=token)}
+
+        # aiohttps does not allow to send auth argument and auth header together
+        self.session._default_auth = None
+
+        async with self.session.get(self.url, headers=headers, proxy=self.proxy) as response:
+            response.raise_for_status()
+            to_return = await self._handle_response(response)
+            await response.release()
+
+        if self._close_session_on_finalize:
+            self.session.close()
+        return to_return
```

Comment:
You could call super() to receive all of this instead.

Reply:
I couldn't, pulpcore is different:
```python
async with self.session.get(self.url, proxy=self.proxy, auth=self.auth) as response:
```
https://github.com/pulp/pulpcore/blob/master/pulpcore/download/http.py#L197